server/src/main/java/org/opensearch/common/cache/store/EhCacheDiskCache.java

```diff
@@ -0,0 +1,543 @@
+import org.ehcache.config.builders.ResourcePoolsBuilder;
+import org.ehcache.config.units.MemoryUnit;
+import org.ehcache.event.CacheEvent;
+import org.ehcache.event.CacheEventListener;
+import org.ehcache.event.EventType;
+import org.ehcache.expiry.ExpiryPolicy;
+import org.ehcache.impl.config.store.disk.OffHeapDiskStoreConfiguration;
+import org.ehcache.spi.loaderwriter.CacheLoadingException;
+import org.ehcache.spi.loaderwriter.CacheWritingException;
+
+/**
+ * This variant of disk cache uses Ehcache underneath.
+ *  @param <K> Type of key.
+ *  @param <V> Type of value.
+ *
+ *  @opensearch.experimental
+ *
+ */
+public class EhCacheDiskCache<K, V> implements StoreAwareCache<K, V> {
+
+    private static final Logger logger = LogManager.getLogger(EhCacheDiskCache.class);
+
+    // A Cache manager can create many caches.
+    private final PersistentCacheManager cacheManager;
+
+    // Disk cache
+    private Cache<K, V> cache;
+    private final long maxWeightInBytes;
+    private final String storagePath;
+
+    private final Class<K> keyType;
+
+    private final Class<V> valueType;
+
+    private final TimeValue expireAfterAccess;
+
+    private final DiskCacheStats stats = new DiskCacheStats();
+
+    private final EhCacheEventListener<K, V> ehCacheEventListener;
+
+    private final String threadPoolAlias;
+
+    private final Settings settings;
+
+    private final static String DISK_CACHE_ALIAS = "ehDiskCache";
+
+    private final static String THREAD_POOL_ALIAS_PREFIX = "ehcachePool";
+
+    private final static int MINIMUM_MAX_SIZE_IN_BYTES = 1024 * 100; // 100KB
+
+    // Ehcache disk write minimum threads for its pool
+    public final Setting<Integer> DISK_WRITE_MINIMUM_THREADS;
+
+    // Ehcache disk write maximum threads for its pool
+    public final Setting<Integer> DISK_WRITE_MAXIMUM_THREADS;
+
+    // Not be to confused with number of disk segments, this is different. Defines
+    // distinct write queues created for disk store where a group of segments share a write queue. This is
+    // implemented with ehcache using a partitioned thread pool exectutor By default all segments share a single write
+    // queue ie write concurrency is 1. Check OffHeapDiskStoreConfiguration and DiskWriteThreadPool.
+    public final Setting<Integer> DISK_WRITE_CONCURRENCY;
+
+    // Defines how many segments the disk cache is separated into. Higher number achieves greater concurrency but
+    // will hold that many file pointers. Default is 16.
+    public final Setting<Integer> DISK_SEGMENTS;
+
+    private final StoreAwareCacheEventListener<K, V> eventListener;
+
+    /**
+     * Used in computeIfAbsent to synchronize loading of a given key. This is needed as ehcache doesn't provide a
+     * computeIfAbsent method.
+     */
+    Map<K, CompletableFuture<Tuple<K, V>>> completableFutureMap = new ConcurrentHashMap<>();
+
+    private EhCacheDiskCache(Builder<K, V> builder) {
```

Comment:
This class has some serious complexity, lets dedicate some time to build up unit tests to validate the behavior/assumptions about how these systems work in addition to the scenario tests

Reply:
I think I have added sufficient unit tests to validate a lot of scenarios, let me know if you have something specific.
Plus this is a part of larger tiered cache framework and yet to be integrated fully. Those will have additional tests coming up as part of upcoming integration PRs after this.

Reply:
I don't see the code coverage report - that would be my guide to feel confidant we had enough test coverage, maybe it isn't being generated because of the check failures, can you look into that?

Reply:
Looks like we're at:



.../opensearch/cache/store/disk/EhcacheDiskCache.java | 75.00% | 38 Missing and 9 partials
-- | -- | --


[.../opensearch/cache/store/disk/EhcacheDiskCache.java](https://app.codecov.io/gh/opensearch-project/OpenSearch/pull/11874?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opensearch-project#diff-cGx1Z2lucy9jYWNoZS1laGNhY2hlL3NyYy9tYWluL2phdmEvb3JnL29wZW5zZWFyY2gvY2FjaGUvc3RvcmUvZGlzay9FaGNhY2hlRGlza0NhY2hlLmphdmE=)	75.00%	38 Missing and 9 partials

Checking [the report](https://app.codecov.io/gh/opensearch-project/OpenSearch/pull/11874/blob/plugins/cache-ehcache/src/main/java/org/opensearch/cache/store/disk/EhcacheDiskCache.java), it looks like most of the missing coverage is on `catch` clauses and other failure cases.

I think we're pretty good. Also, since we've moved it our of core, into an opt-in plugin, I feel like 75% is a decent coverage number.

Reply:
Thanks for checking in @msfroh.  For a plugin marked with experimental annotation, I'm fine with this.  As we invest more in this space and we have users that can be broken writing tests to make sure we keep supporting them will be important- but shouldn't block the great progress in this PR.  - I'm resolving this thread and will merge this change.